[FIX] web: Display correct Favicon depending company

nboulif opened the pull request:
Steps to reproduce:

 - Install the module web_enterprise
 - Activate Debug mode
 - Go to Settings -> Companies
 - Create company A and B
 - Upload new **pictures** and use them as Favicon for each company
 - Refresh the page or switch company

Issue:

  The favicon is not the right one.

Cause:

  When starting the webclient, the main favicon in the header is updated
  (with the correct current company id from cookies) in JS.
  Since the current company favicon is a picture, the browser fallback
  on the next favicon available (not the first one since the favicon is
  a picture), who, if web_enterprise is installed, is the static mobile
  (android) icon.

Solution:

  Update all icons with JS so it will fallback on the first declared icon
  (even though it's a picture) and also fix the issue of having a static
  mobile icon in enterprise.

opw-2664525

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/80113).

nboulif commented:
Enterprise PR: https://github.com/odoo/enterprise/pull/22411

nboulif commented:
Hello @rdeodoo :slightly_smiling_face: 
I have a question regarding my fix:
I tried to find if the selected companies are available in the template as variable (or in the context or request), but there are not...
Therefore, I added the 'cids' from the cookies in `_prepare_qcontext`, however, I'm pretty sure it's not the right place :sweat_smile: 
any clue? Thanks in advance

rdeodoo commented on odoo/addons/base/models/ir_ui_view.py:
```diff
@@ -1203,9 +1203,14 @@ def _prepare_qcontext(self):
             to render website layout template)
         """
+        user_id = self.env["res.users"].browse(self.env.user.id)
+        cids = [str(user_id.company_id.id)]
```
`cids` gives no clue of what it holds :-)
If this were to stay, rename it `company_ids`

rdeodoo replied:
`cids` back to life :D
We usually use incomprehensible and very short cookies name to lighter the data send from the client, as the cookies are sent on every http request and there is no cache mechanism, so any gain is good.
But for a variable name, it really should be avoided :) `cids` doesn't mean anything

nboulif replied:
ahah my apologies, I forget to answer you. At first, I wanted to match this https://github.com/odoo/odoo/blob/4c1271738daa039e79a11646fef3f4264e4d96d2/addons/mail/controllers/main.py#L77 
Thanks for the clarification!

rdeodoo reviewed (commented):
As this is a pure webclient issue/question, may I suggest you check with them? They will definitely know more about it.

ged-odoo reviewed (approved):
looks good to me

kebeclibre commented on odoo/addons/base/models/ir_ui_view.py:
```diff
@@ -1203,9 +1203,14 @@ def _prepare_qcontext(self):
         """ Returns the qcontext : rendering context with website specific value (required
             to render website layout template)
         """
+        user_id = self.env["res.users"].browse(self.env.user.id)
```
This method is used by anyone tryng to render a view. I don't think we wanna push those info at this point.

Rather, look into `webclient_rendering_context` (or the v13.0 equivalent if different) to render the webclient_bootstrap template with the right icon

the website module already has that kind of mechanism
`        <t t-set="x_icon" t-value="website.image_url(website, 'favicon')"/>`

nboulif commented:
@gde-odoo  @kebeclibre Thanks for review!
I updated the fix, however not sure it's the best way, I thought maybe altering the `curent_company` value:
```python
current_company = user.company_id
if request and hasattr(request, 'httprequest'):
    cids = request.httprequest.cookies.get('cids')
    current_company = (cids and user.company_ids.filtered(lambda x: x.id == int(cids.split(",")[0]))) or current_company
session_info.update({
    # current_company should be default_company if not set in cookies
    "user_companies": {
        'current_company': (current_company.id, current_company.name), 
        'allowed_companies': [(comp.id, comp.name) for comp in user.company_ids],
    },
````
and do in template

odony commented on addons/web/models/ir_http.py:
```diff
@@ -62,9 +62,16 @@ def session_info(self):
+                "user_companies": {
+                    'current_company': (user.company_id.id, user.company_id.name),
+                    'allowed_companies': [(comp.id, comp.name) for comp in user.company_ids],
+                    "company_ids": company_ids,
```
In general you shouldn't be modifying `ir_http.py`, `http.py` or any other low-level server-side component unless you have a very good reason to ;-)

Here, it's not clear to me why that change is necessary, and worse, it seems to be introducing confusing notions of what companies are currently enabled on the server-side: the `allowed_companies` value may be different from your new `company_ids` value.

If I remember correctly our design [1], it's the client-side who is responsible for parsing the `cids` and determining the appropriate `allowed_company_ids` value to pass in the context. And there are already several places where it's doing that, so there should be no need to copy that logic server-side.


[1] See a5b6f31cf28e5381e1c85f66730bcdb55998e643

JKE-be reviewed (commented):
Just tested in v13 and v14; and I cannot reproduce your problem.
On v15, it doesn't seems to work... but it could be explain by the commit https://github.com/odoo/odoo/commit/0573acae2306bf5da2005852da9323ddc59e5431 that remove (or rewrite?) the feature no ? 

https://drive.google.com/file/d/1Mon7r8pRqP9mSc4vFv08y5W3rcolBikp/view

So what are your trying to fix in the v13?

nboulif commented:
Hello @JKE-be , thanks for the check!
In fact, the issue is present only with enterprise module, especially as soon as `web_enterprise` is installed:
`web_enterprise` will update the template `web.webclient_bootstrap` by adding some icon for mobiles in the header:
https://github.com/odoo/enterprise/blob/cff73ea89cb16668b5402300bdf7150e25535f34/web_enterprise/views/webclient_templates.xml#L123-L140
By doing that, it will bring the real issue: 
https://github.com/odoo/odoo/blob/6a2cc6b2cd4c3a345207796a948d6f5a2c2f68da/addons/web/static/src/js/chrome/abstract_web_client.js#L135
When updating manually `link[type='image/x-icon']` in JS, chrome will take the the next available icon declared in the header, in this case:
https://github.com/odoo/enterprise/blob/cff73ea89cb16668b5402300bdf7150e25535f34/web_enterprise/views/webclient_templates.xml#L127
Therefore, it will always display the wrong favicon in enterprise.

My fix has as purpose to get the current activated company (`cids` from cookies if set) and give it to the template directly instead of re-writing it in the JS.

Another fix (that I think is ugly), would be to remove in JS all icons then add them back with main icon in specific order...

JKE-be commented:
Why not just update all icons in js in one shot? 

Smthg like `$("link[rel*='icon']").attr('href', '/web/image/res.company/' + String(stateCompanyIDS[0]) + '/favicon/')`;

nboulif commented:
@JKE-be It is indeed working but wasn't sure it's the best way... Glad you put your input :pray: 
I updated the PR :+1: 
PS: Do you think I should also update the Windows Phone icon too, so all have accurate value?
`$("meta[name='msapplication-TileImage']").attr('content', '/web/image/res.company/' + String(stateCompanyIDS[0]) + '/favicon/');
`

rdeodoo commented:
Just my 2cents reading the notification emails from github (was curious)

@nboulif https://github.com/odoo/odoo/pull/80113#issuecomment-1026208768 is a nice and detailed explanation, which is IMHO laking inside the commit message*. Proof is, I guess, that JKE did not understand what you were trying to fix or what was the actual issue, while it seems like your explanation provided everything that was needed in order to figure things.

I'd personally add a link to that comment in your commit message like `Find detailed explanation here: xxx` or try to summarize it and put it inside the commit message. Sad to see such detailed info that seems needed to understand the issue*, on the PR but not in the commit :)

*Especially parts like this `the issue is present only with enterprise module, especially as soon as web_enterprise is installed`

nboulif commented:
@rdeodoo updated

JKE-be commented:
> PS: Do you think I should also update the Windows Phone icon too, so all have accurate value?
$("meta[name='msapplication-TileImage']").attr('content', '/web/image/res.company/' + String(stateCompanyIDS[0]) + '/favicon/');

yes, ofc. Why would you like the feature on Mac but not on Windows. As well we do it for all, as well for none... make sense no?

nboulif commented:
> yes, ofc. Why would you like the feature on Mac but not on Windows. As well we do it for all, as well for none... make sense no?

Stupid question (seems just weird to adapt explicitly window phone only icon in community), make sense indeed. Updated :+1:

JKE-be reviewed (commented):
Seems better from this way - shorter ;) Thank for the update.
Details but the trailing / is useless and could trigger a useless redirect.

But I still have a problem, I cannot reproduce your problem.
Without your changes, If I take the last enterprise v13/14 on runbot (with demo data and so multi company multi favicon) all works as expected.

I'm pretty sure that your commit will fix the case for some device / os. But the explanation is missing in your commit/pr msg; and if I just read your commit msg, this diff seems useless since I have no bug with chrome on Ubuntu.

nboulif commented:
@JKE-be 
> Details but the trailing / is useless and could trigger a useless redirect.

Since was already there in JS, wasn't sure if I can remove it, you confirmed it :+1: 

> But I still have a problem, I cannot reproduce your problem.
Without your changes, If I take the last enterprise v13/14 on runbot (with demo data and so multi company multi favicon) all works as expected.

Sincere apologize, I forget to specify in the commit with the steps to reproduce that must change the default favicon with a **picture** (and not a **favicon**).
It's an important remark since, therefore, not sure finally if there is a bug or just a limitation when uploading a picture...

> I'm pretty sure that your commit will fix the case for some device / os. But the explanation is missing in your commit/pr msg; and if I just read your commit msg, this diff seems useless since I have no bug with chrome on Ubuntu.

Indeed, I will update the commit message (if still considered as a bug), by adding the step `change favicon with a **picture**` and updating the solution message that it will prevent having an undefined behavior when using a favicon as picture and also override the static mobile icon from enterprise, right?

Of course, it will be another commit message for v15.0. 

> Remark: from what I read in the opw, the customer doesn't complains about this one (and so this fix is maybe even useless for the opw). He is using chrome. As specified by GOR it is broken in v15 only, and he report a problem in v14. So you should double check

Since I tried to upload a picture instead of a favicon, I did get the undefined behavior that was similar to the ticket issue and therefore started investigating on the oldest supported version where I was able to reproduce 'the issue', not an excuse.

> Remark2: some tickets are closed in v15 about this favicon (opw2733691 e.g.), under pretext that you already fix it. But be aware, than this diff is until v15. After the feature is really broken and need to be fixed!

I guess the feature has been removed during the refractoring. My thought was to do the same fix (with allowedCompanyIds) here:
https://github.com/odoo/odoo/blob/54e9527ec2fdf0467de08b372c7e891fedaa68ae/addons/web/static/src/webclient/company_service.js#L48

rdeodoo commented:
> In fact, the issue is present only with enterprise module, especially as soon as web_enterprise is installed:

> Sincere apologize, I forget to specify in the commit with the steps to reproduce that must change the default favicon with a picture (and not a favicon).

Sounds like a fun game :')

JKE-be commented:
> Sincere apologize, I forget to specify in the commit with the steps to reproduce that must change the default favicon with a picture (and not a favicon).
It's an important remark since, therefore, not sure finally if it is a bug or just a limitation when uploading a picture... (still fix enterprise)

I think that you reply to yourself...
Yes the problem is that we expect a favicon and not an img.

Since v13 is a stable version and that an alternative exists for the opw: use an ico file directly, you should not risk to break the stable version and keep it as this.

But add in master the same mimic than _handle_favicon in website to improve the UX.

And you can apply this patch only in v15 to support again the custom favicon.

JKE-be commented on addons/web/static/src/webclient/webclient.js:
```diff
@@ -21,6 +21,7 @@ export class WebClient extends Component {
         this.title = useService("title");
         this.router = useService("router");
         this.user = useService("user");
+        this.companyService = useService("company");
```
```suggestion
        this.company = useService("company");
```

nboulif replied:
Done :+1:

JKE-be commented on addons/web/static/src/webclient/webclient.js:
```diff
@@ -87,6 +88,8 @@ export class WebClient extends Component {
             // If no action => falls back to the default app
             await this._loadDefaultApp();
         }
+        $("link[rel*='icon']").attr('href', '/web/image/res.company/' + String(this.companyService.currentCompany.id) + '/favicon');
+        $("meta[name='msapplication-TileImage']").attr('content', '/web/image/res.company/' + String(this.companyService.currentCompany.id) + '/favicon');
```
```suggestion
        const favicon = `/web/image/res.company/${this.company.currentCompany.id}/favicon`;
        $("link[rel*='icon']").attr('href', favicon);
        $("meta[name='msapplication-TileImage']").attr('content', favicon);
```

nboulif commented:
@JKE-be Updated. Thanks again!

JKE-be commented:
but not sure about the lifecycle of owl and if into the right place.
maybe @aab-odoo can better help you about the right place and why your test is wrong

nboulif commented:
@JKE-be I updated the fix: the test failed because the `company` service was missing in the helpers.

aab-odoo commented on addons/web/static/src/webclient/webclient.js:
```diff
@@ -87,6 +88,9 @@ export class WebClient extends Component {
         }
+        const favicon = `/web/image/res.company/${this.company.currentCompany.id}/favicon`;
+        $("link[rel*='icon']").attr('href', favicon);
+        $("meta[name='msapplication-TileImage']").attr('content', favicon);
```
Note that generating the page with the correct icon directly would still be the best solution imho, but I understand the arguments that the cids logic is implemented client side, so I guess it's the best compromise. Maybe the location of this cids logic is a problem though.

nboulif replied:
Thanks for the review. Updated :+1:

JKE-be commented:
@robodoo r+

:+1: Thanks
